smart_contracts/vm2/macros/src/lib.rs

```diff
@@ -1320,7 +1357,48 @@ fn casper_trait_definition(mut item_trait: ItemTrait, trait_meta: TraitMeta) ->
                             "can't make dispatcher for private static method"
                         );
                         quote! {
-                            #vis extern "C"  fn #dispatch_func_name<T: #trait_name>() {
+                            #[cfg(not(target_arch = "wasm32"))]
+                            #vis fn #dispatch_func_name<T: #trait_name>() {
+                                #[derive(#crate_path::serializers::borsh::BorshDeserialize)]
+                                #[borsh(crate = #borsh_path)]
+                                struct Arguments {
+                                    #(#args_attrs,)*
+                                }
+
+                                let mut flags = #crate_path::common::flags::ReturnFlags::empty();
+                                let input = #crate_path::prelude::casper::copy_input();
+                                let args: Arguments = {
+                                    match #resolve_abi_convention {
+                                        casper_contract_sdk::serializers::AbiConvention::Positional => {
+                                            casper_contract_sdk::serializers::borsh::from_slice(&input).unwrap()
+                                        }
+                                        casper_contract_sdk::serializers::AbiConvention::Named => {
+                                            let runtime_args: casper_contract_sdk::compat::types::RuntimeArgs =
+                                                casper_contract_sdk::serializers::borsh::from_slice(&input).unwrap();
+                                                #(
+                                                    let #arg_names: #arg_types = {
+                                                        let cl_value = runtime_args.get(stringify!(#arg_names)).unwrap_or_else(|| panic!(concat!("Failed to get named argument \"", stringify!(#arg_names), "\"")));
+                                                        cl_value.to_t::<#arg_types>().unwrap_or_else(|error| {
+                                                            panic!(concat!("Failed to convert named argument \"", stringify!(#arg_names), "\": {}"), error)
+                                                        })
+                                                    };
+                                                )*
+
+                                            Arguments {
+                                                #(
+                                                    #arg_names,
+                                                )*
+                                            }
+                                        }
+                                    }
+                                };
+                                let _ret = <T as #trait_name>::#func_name(#(args.#arg_names,)*);
+
+                                #handle_ret
+                            }
+
+                            #[cfg(target_arch = "wasm32")]
```

Comment:
Like above, if we're at it, it's worth checking if we really need two copies of `extern "C"` generic function